common/src/main/java/com/wynntils/wynn/model/ShamanTotemModel.java

```diff
@@ -0,0 +1,301 @@
+        }
+    }
+
+    @SubscribeEvent
+    public void onTotemDestroy(RemoveEntitiesEvent e) {
+        if (!WynnUtils.onWorld()) return;
+
+        List<Integer> destroyedEntities = e.getEntityIds();
+
+        if (totem1 != null && destroyedEntities.contains(totem1.getTimerId())) {
+            removeTotem(1);
+        }
+        if (totem2 != null && destroyedEntities.contains(totem2.getTimerId())) {
+            removeTotem(2);
+        }
+        if (totem3 != null && destroyedEntities.contains(totem3.getTimerId())) {
+            removeTotem(3);
+        }
+    }
+
+    @SubscribeEvent
+    public void onClassChange(CharacterUpdateEvent e) {
+        removeAllTotems();
+    }
+
+    @SubscribeEvent
+    public void onHeldItemChange(ChangeCarriedItemEvent e) {
+        removeAllTotems();
+    }
+
+    private String getNameFromMetadata(List<SynchedEntityData.DataValue<?>> data) {
+        // The rename stuff we're looking for is eventually something like
+        // Optional[literal{§c26s}[style={}]]
+        for (SynchedEntityData.DataValue<?> packedItem : data) {
+            if (!(packedItem.value() instanceof Optional<?> packetData)
+                    || packetData.isEmpty()
+                    || (!(packetData.get() instanceof MutableComponent content))) continue;
+            return content.toString();
+        }
+        return null;
+    }
+
+    private Entity getBufferedEntity(int entityId) {
+        Entity entity = McUtils.mc().level.getEntity(entityId);
+        if (entity != null) return entity;
+
+        if (entityId == -1) {
+            return new net.minecraft.world.entity.decoration.ArmorStand(McUtils.mc().level, 0, 0, 0);
+        }
+
+        return null;
+    }
+
+    /**
+     * Removes the given totem from the list of totems.
+     * @param totem The totem to remove. Must be 1 or 2.
+     */
+    private void removeTotem(int totem) {
+        switch (totem) {
+            case 1 -> {
+                WynntilsMod.postEvent(new TotemRemovedEvent(1, totem1));
+                totem1 = null;
+                pendingTotem1Id = null;
+                return;
+            }
+            case 2 -> {
+                WynntilsMod.postEvent(new TotemRemovedEvent(2, totem2));
+                totem2 = null;
+                pendingTotem2Id = null;
+                return;
+            }
+            case 3 -> {
+                WynntilsMod.postEvent(new TotemRemovedEvent(3, totem3));
+                totem3 = null;
+                pendingTotem3Id = null;
+                return;
+            }
+            default -> throw new IllegalArgumentException("Totem must be 1, 2, or 3");
+        }
+    }
+
+    /**
+     * Resets all three totem variables.
+     */
+    private void removeAllTotems() {
+        removeTotem(1);
+        removeTotem(2);
+        removeTotem(3);
+        nextTotemSlot = 1;
+    }
+
+    private int getNextTotemSlot() {
+        if (nextTotemSlot == 3) return 1;
+        return nextTotemSlot += 1;
```

Comment:
Are you always alternating between these numbers? What if you cast one totem, wait until it disappears, and cast another. Will the first get number 1 and the second number 2? I might be missing something here...

Reply:
Totems will always fill the slot with the lowest number. So if one totem is cast, it will be in slot 1. Then it expires frees up slot 1. Then another one is cast and it will be in slot 1 again. If two are cast together, they will fill 1 and 2.

Reply:
Perfect, that is what I wanted to hear. :) So the wraparound above is more of a defensive programming, since you can really never have more than 3 totems?